pkg/asset/manifests/ingress.go

```diff
@@ -37,38 +41,81 @@ func (*Ingress) Dependencies() []asset.Asset {
 	}
 }
 
-// Generate generates the ingress config and its CRD.
+// Generate generates the ingress cluster config and default ingresscontroller.
+//
+// A cluster ingress config is always created.
+//
+// A default ingresscontroller is only created if the cluster is using an internal
+// publishing strategy. In this case, the default ingresscontroller is also set
+// to use the internal publishing strategy.
 func (ing *Ingress) Generate(dependencies asset.Parents) error {
 	installConfig := &installconfig.InstallConfig{}
 	dependencies.Get(installConfig)
 
-	config := &configv1.Ingress{
+	ing.FileList = []*asset.File{}
+
+	clusterConfig, err := ing.generateClusterConfig(installConfig.Config)
+	if err != nil {
+		return errors.Wrap(err, "failed to create cluster config")
+	}
+	ing.FileList = append(ing.FileList, &asset.File{
+		Filename: clusterIngressConfigFile,
+		Data:     clusterConfig,
+	})
+
+	defaultIngressController, err := ing.generateDefaultIngressController(installConfig.Config)
+	if err != nil {
+		return errors.Wrap(err, "failed to create default ingresscontroller")
+	}
+	if len(defaultIngressController) > 0 {
+		ing.FileList = append(ing.FileList, &asset.File{
+			Filename: defaultIngressControllerFile,
+			Data:     defaultIngressController,
+		})
+	}
+
+	return nil
+}
+
+func (ing *Ingress) generateClusterConfig(config *types.InstallConfig) ([]byte, error) {
+	obj := &configv1.Ingress{
 		TypeMeta: metav1.TypeMeta{
-			APIVersion: configv1.SchemeGroupVersion.String(),
+			APIVersion: configv1.GroupVersion.String(),
 			Kind:       "Ingress",
 		},
 		ObjectMeta: metav1.ObjectMeta{
 			Name: "cluster",
 			// not namespaced
 		},
 		Spec: configv1.IngressSpec{
-			Domain: fmt.Sprintf("apps.%s", installConfig.Config.ClusterDomain()),
+			Domain: fmt.Sprintf("apps.%s", config.ClusterDomain()),
 		},
 	}
+	return yaml.Marshal(obj)
+}
 
-	configData, err := yaml.Marshal(config)
-	if err != nil {
-		return errors.Wrapf(err, "failed to create %s manifests from InstallConfig", ing.Name())
-	}
-
-	ing.FileList = []*asset.File{
-		{
-			Filename: ingCfgFilename,
-			Data:     configData,
-		},
+func (ing *Ingress) generateDefaultIngressController(config *types.InstallConfig) ([]byte, error) {
+	switch config.Publish {
+	case types.InternalPublishingStrategy:
+		obj := &operatorv1.IngressController{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: operatorv1.GroupVersion.String(),
+				Kind:       "IngressController",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: "openshift-ingress-operator",
+				Name:      "default",
+			},
+			Spec: operatorv1.IngressControllerSpec{
+				EndpointPublishingStrategy: &operatorv1.EndpointPublishingStrategy{
+					Type: operatorv1.PrivateStrategyType,
+				},
+			},
+		}
+		return yaml.Marshal(obj)
+	default:
+		return []byte{}, nil
```

Comment:
should we return `nil, nil`, why return a byte slice of `len 0`

Reply:
Fixed